src/com/amazon/corretto/crypto/provider/AesGcmSpi.java

```diff
@@ -924,6 +981,10 @@ private void checkNeedReset() {
 
     // @GuardedBy("this") // Restore once replacement for JSR-305 available
     private void stateReset() {
+        // While this shouldn't happen, we cover this case to ensure we return to a good state.
+        if (context != null && context.isReleased()) {
+            context = null;
+        }
         decryptInputBuf.reset();
         decryptAADBuf.reset();
```

Comment:
not particularly pertinent to this change, but do we define a max size for our AccessibleByteArrayOutputStream?

if so, and the max is reasonably small, would it be advantageous for performance to back that with a stack-allocated array, and then just maintain a "high watermark" value to track the current size of the buffer (or similarly, track start/end indices and use it like a ring buffer)? way we could avoid allocations on resize as well as the cost of `Arrays.fill(0)` on calls to `reset()` -- just set the watermark or head/tail to 0. i suppose we wouldn't want to leak e.g. plaintext from prior invocations of the cipher, but we could prevent that by ensuring we don't expose the buffer reference anywhere.

Reply:
It might be possible, but we don't have a max size in general (some cases we do). I'm not sure how useful a stack allocated array would be. We already maintain a high water mark and thus avoid extra resizes.

There is an interesting experiment I made in a private branch where I implemented an [ArrayCache](https://github.com/SalusaSecondus/amazon-corretto-crypto-provider/blob/buff_pool/src/com/amazon/corretto/crypto/provider/ArrayCache.java) to specifically reuse arrays of common sizes to avoid allocation/GC costs. While I think it is probably useful in the grand scheme of things, it wasn't sufficiently valuable *then* (as there were bigger/easier improvements to be had).